File: 백준/Gold/8980. 택배/택배.py
import sys

# 입력 받기
N, C = map(int, sys.stdin.readline().split())
M = int(sys.stdin.readline())

tackBaes = []
for _ in range(M):
    start, end, amount = map(int, sys.stdin.readline().split())
    tackBaes.append([start, end, amount])

# 정렬: 도착 마을 기준 내림차순, 같으면 출발 마을 내림차순
tackBaes.sort(key=lambda x: (-x[1], -x[0]))

# 마을별 적재량 배열
arr = [0] * (N + 1)
ans = 0

# 택배 배송 처리
for start, end, amount in tackBaes:
    maxC = C

    for i in range(start + 1, end + 1):
        maxC = min(maxC, C - arr[i])

    can = min(amount, maxC)  # 적재 가능한 박스 개수 결정

    for i in range(start + 1, end + 1):
        arr[i] += can

    ans += can

print(ans)

File: 백준/Gold/8980. 택배/택배.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int start, end, amount;
} TackBae;

// 비교 함수 (도착 마을 기준 정렬, 같으면 출발 마을 오름차순)
int compare(const void *a, const void *b) {
    TackBae *t1 = (TackBae*)a;
    TackBae *t2 = (TackBae*)b;
    if (t1->end == t2->end) {
        return t1->start - t2->start;
    }
    return t1->end - t2->end;
}

int main() {
    int N, C, M;
    scanf("%d %d", &N, &C);
    scanf("%d", &M);

    TackBae tackBaes[M];
    int arr[N+1]; // 각 마을 간의 트럭 적재량을 저장할 배열
    for (int i = 0; i <= N; i++) arr[i] = 0; // 초기화

    // 입력 받기
    for (int i = 0; i < M; i++) {
        scanf("%d %d %d", &tackBaes[i].start, &tackBaes[i].end, &tackBaes[i].amount);
    }

    // 정렬 수행
    qsort(tackBaes, M, sizeof(TackBae), compare);

    int ans = 0;
    // 각 배송 요청 처리
    for (int i = 0; i < M; i++) {
        int maxC = C;
        
        // 구간 내 최대 적재 가능 용량 계산
        for (int j = tackBaes[i].start; j < tackBaes[i].end; j++) {
            if (C - arr[j] < maxC) {
                maxC = C - arr[j];
            }
        }

        int can = (tackBaes[i].amount < maxC) ? tackBaes[i].amount : maxC;

        // 구간 업데이트
        for (int j = tackBaes[i].start; j < tackBaes[i].end; j++) {
            arr[j] += can;
        }

        ans += can;
    }

    printf("%d\n", ans);
    return 0;
}
